### PR TITLE
2231 conan compiler detection

### LINKED ISSUE #2231: Conan on Mac can't recognize compiler correctly
Hello, 
https://travis-ci.org/procxx/kepka/jobs/322618396

on the link you can see that Conan recognize compiler as gcc 4.2 on Travis, but it's not true.

### PR DESCRIPTION
**Pull request by pvicente**
- [X] Refer to the issue that supports this Pull Request.
- [X] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [X] I've followed the PEP8 style guides for Python code.

closes #2231 

Previously failing compiler detection in mac os x when gcc was forced:

```bash
09:28 $ CC=gcc CXX=g++ conan install boost/1.66.0@conan/stable
Auto detecting your dev setup to initialize the default profile (/Users/pvicente/.conan/profiles/default)
CC and CXX: gcc, g++
Found gcc 4.2
Default settings
	os=Macos
	os_build=Macos
	arch=x86_64
	arch_build=x86_64
	compiler=gcc
	compiler.version=4.2
	compiler.libcxx=libstdc++
	build_type=Release
*** You can change them in /Users/pvicente/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***


ERROR: Invalid setting '4.2' is not a valid 'settings.compiler.version' value.
Possible values are ['4.1', '4.4', '4.5', '4.6', '4.7', '4.8', '4.9', '5', '5.1', '5.2', '5.3', '5.4', '5.5', '6', '6.1', '6.2', '6.3', '6.4', '7', '7.1', '7.2', '7.3']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"
```

Now:

```bash
09:26 $ CC=gcc CXX=g++ conan install boost/1.66.0@conan/stable
Auto detecting your dev setup to initialize the default profile (/Users/pvicente/.conan/profiles/default)
CC and CXX: gcc, g++
Found apple-clang 9.0
Default settings
	os=Macos
	os_build=Macos
	arch=x86_64
	arch_build=x86_64
	compiler=apple-clang
	compiler.version=9.0
	compiler.libcxx=libc++
	build_type=Release
*** You can change them in /Users/pvicente/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***
...
```

### CODE REVIEW

File: conans/client/conf/detect.py [hunk 1]
```diff
@@ -146,9 +159,9 @@ def _get_default_compiler(output):
     if cc or cxx:  # Env defined, use them
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
-        if "gcc" in command:
+        if _is_real_compiler("gcc", command):
```
Comment (lasote) [inline conans/client/conf/detect.py]: I think it is clearer to get here the output of the "command --help", the "if "gcc" in output:" and "if "clang" in output". The function is real compiler feels a little confusing. Maybe also check the output in lowercase in case GCC or Clang in output.

Comment (lasote) [reply conans/client/conf/detect.py]: Sorry for my comments formatting. I'm with the mobile.

Comment (memsharded) [reply conans/client/conf/detect.py]: Yes, agree, I'd think it would be more understandable. I think that @lasote meant ``command --version``, not ``--help``.

File: conans/test/util/detect_test.py [hunk 2]
```diff
@@ -1,15 +1,36 @@
...
+        result = dict(result)
+        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
+        if platform_compiler is not None:
+            self.assertEquals(result.get("compiler", None), platform_compiler)
+
+    @unittest.skipIf(platform.system() != "Darwin", "Test only running in Mac OS X")
+    def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        # See: https://github.com/conan-io/conan/issues/2231
+        output = TestBufferConanOutput()
+        with tools.environment_append({"CC": "gcc"}):
```
Comment (memsharded) [inline conans/test/util/detect_test.py]: This test is nice, but is not testing the reported issue. Which is gcc being actually a symlink to clang. It might be difficult to fully automate, so I am fine if tested manually.

**Comment by SSE4** [comment]
may be it's better and more reliable to avoid parsing `--version` output at all, and use instead macro definitions reported by compiler as shown at https://github.com/conan-io/conan/issues/1127?

File: conans/client/conf/detect.py (hunk 1 above)
Comment (pvicente) [reply conans/client/conf/detect.py]: @lasote  do you mean something like this?

```python
if "gcc" in command:
    _, out = _execute("%s --version", command)
    if "gcc" in out.lower():
         return _gcc_compiler(output, command)
if "clang" in command.lower():
   _, out =  _execute("%s --version", command.lower())
   if "clang" in out.lower():
        return _clang_compiler(output, command)   
```

File: conans/test/util/detect_test.py (hunk 2 above)
Comment (pvicente) [reply conans/test/util/detect_test.py]: Indeed, this test is only executed in MacOS x (Darwin) where gcc command is clang instead clang which was the bug report for this issue.

See output:
```
17:06 $ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
and the test is checking that detected compiler using gcc should be ``apple-clang`` https://github.com/conan-io/conan/pull/2487/files#diff-32d9d50eb3e70bdef7cfb86080926414R13

File: conans/client/conf/detect.py (hunk 1 above)
Comment (pvicente) [reply conans/client/conf/detect.py]: I've fixed it with 2nd option

File: conans/client/conf/detect.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def _clang_compiler(output, compiler_exe="clang"):
             compiler = "apple-clang"
         elif "clang version" in out:
             compiler = "clang"
-        installed_version = re.search("([0-9]\.[0-9])", out).group()
+        installed_version = re.search("version ([0-9]\.[0-9])", out).group(1)
```
Comment (lasote) [inline conans/client/conf/detect.py]: This change in the regular expression worries me a little. If some specific clang is not printing the "version" word we could be breaking.  Why did you need it? If there is a good reason I'm ok with it, if don't, better keep it like it was.

**Comment by lasote** [comment]
@SSE4 could be more accurate, but I think we are good by now with the `--version` until we detect that it is not enough. Anyway thanks for your suggestion!

File: conans/client/conf/detect.py (hunk 3 above)
Comment (lasote) [reply conans/client/conf/detect.py]: If it is only for apple-clang I would prefer to limit it to the Apple if to mitigate possible issues

Comment (memsharded) [reply conans/client/conf/detect.py]: But apple-clang was already working, when the compiler is "clang --version". So parsing didn't seem an issue

File: conans/test/util/detect_test.py (hunk 2 above)
Comment (memsharded) [reply conans/test/util/detect_test.py]: It is not to make it more evident. It could happen that the CI machine could have *gcc* really installed in it, and the test would pass without checking for clang output, so it wouldn't test at all the reported bug fix. It is necessary to explicitly check for "apple-clang" in the output

File: conans/client/conf/detect.py (hunk 3 above)
Comment (lasote) [reply conans/client/conf/detect.py]: That's a good question, but looking at the output it could be:

```
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.4.0
```

Please @pvicente confirm then it is failing and why didn't fail so far.

File: conans/test/util/detect_test.py (hunk 2 above)
Comment (pvicente) [reply conans/test/util/detect_test.py]: @memsharded I don't understand what's the problem? @lasote do you know what are we talking about? This test is only executed in Mac OS X and the CI machine for Mac OS X has fake ``gcc`` as ``apple-clang``, but if you don't want to rely on this assumption maybe it's better to remove this end2end test.

Comment (lasote) [reply conans/test/util/detect_test.py]: Just check the output to confirm that the assumption is true. Because if the assumption change, it is better to know it with a fail test than being happy with a green build and a useless test.

File: conans/client/conf/detect.py (hunk 3 above)
Comment (pvicente) [reply conans/client/conf/detect.py]: The problem is when using fake gcc in mac os x.

This is the output for gcc:

```
10:31 $ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
The previous regular expression catches 4.2 as compiler version, see [regex problem](https://regex101.com/r/cT1W9X/4)

When I added version in regex I checked out that clang in linux, clang in apple are returning ``version`` in the output.

MacOSX: 

```
23:21 $ clang --version
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Linux:

```
conan@8bb2abebb6c8:/Users/pvicente/git/conan/conan$ clang --version
clang version 5.0.0-3 (tags/RELEASE_500/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

File: conans/test/util/detect_test.py (hunk 2 above)
Comment (pvicente) [reply conans/test/util/detect_test.py]: Ok, I got it, thanks for the explanation. I wasn't getting it, sorry guys!!!

**Comment by pvicente** [comment]
@lasote @memsharded fixed test case checking output

File: conans/test/util/detect_test.py [hunk 4]
```diff
@@ -1,15 +1,55 @@
...
-                    self.assertEquals(value, "gcc")
+        # result is a list of tuples (name, value) so converting it to dict
+        result = dict(result)
+        platform_compiler = platform_default_compilers.get(platform.system(), None)
+        if platform_compiler is not None:
+            self.assertEquals(result.get("compiler", None), platform_compiler)
+
+    def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        """
+        In some Mac OS X gcc is in reality clang.
```
Comment (lasote) [inline conans/test/util/detect_test.py]: We investigate it and we changed our mind, the clang "gcc" is not a pure gcc. Read the last comment of @memsharded here: https://github.com/conan-io/conan/issues/2231

Comment (pvicente) [reply conans/test/util/detect_test.py]: @lasote @memsharded I've updated the code to skip gcc detection this case arises as commented in #2231

**Comment by pvicente** [comment]
@lasote @memsharded it's ready for review with suggested changes #2231 

> But I think the logic should probably be: Skip the gcc (as clang frontend) in OSX in the default conan detection of compiler. So in OSX, conan will be able to detect normal "apple-clang" clang compiler, and maybe "gcc", but as complete gcc, not that clang frontend.

File: conans/test/util/detect_test.py [hunk 5]
```diff
@@ -1,15 +1,60 @@
...
+
+    def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        """
+        Test if gcc in Mac OS X is using apple-clang as frontend
+        """
+        # See: https://github.com/conan-io/conan/issues/2231
+        if platform.system() != "Darwin":
+            return
+
+        def _execute(command):
```
Comment (lasote) [inline conans/test/util/detect_test.py]: You can simplify all this with `output_str = subprocess.check_output(command)` and try, except in case of subprocess failure.

Comment (pvicente) [reply conans/test/util/detect_test.py]: Yes, totally right!!! fixing it!!!

File: conans/test/util/detect_test.py [hunk 6]
```diff
@@ -31,23 +31,14 @@ def detect_default_in_mac_os_using_gcc_as_default_test(self):
...
+        except subprocess.CalledProcessError:
+            # gcc is not installed or there is any error (no test scenario)
+            return
 
-        proc_return, output = _execute("gcc --version")
-        if proc_return != 0 or "clang" not in output:
-            # Not test scenario (gcc should display clang in output
+        if "clang" not in output:
+            # Not test scenario gcc should display clang in output
             return
```
Comment (lasote) [inline conans/test/util/detect_test.py]: Mhh, may we raise here to to realize that maybe the default OSX policy for "gcc" has changed?

Comment (lasote) [reply conans/test/util/detect_test.py]: I know that it is not the scope of the test, but useful to know if something changes

Comment (pvicente) [reply conans/test/util/detect_test.py]: Yes It's a good idea

Comment (lasote) [reply conans/test/util/detect_test.py]: like: `raise Exception("Apple gcc doesn't point to clang with gcc frontend anymore! please check")`

File: conans/client/conf/detect.py [hunk 7]
```diff
@@ -147,7 +147,16 @@ def _get_default_compiler(output):
...
+            if platform.system() != "Darwin":
+                return _gcc_compiler(output, command)
+
+            # In Mac OS X check if gcc is a fronted using apple-clang
+            _, out = _execute("%s --version" % command)
+            out = out.lower()
+            if "clang" not in out:
+                return _gcc_compiler(output, command)
+            output.warn("%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command)
+            command = "clang"  # Force use of clang when gcc is detected as a fronted of apple-clang
```
Comment (memsharded) [inline conans/client/conf/detect.py]: I am not sure we want to fallback to clang. Because if the env-var CC and CXX are set to gcc, then, typically CMake will use them when running, and the user might not notice, because it detected "clang" when CC=gcc. What do you think @lasote?

File: conans/client/conf/detect.py [hunk 8]
```diff
@@ -147,7 +147,16 @@ def _get_default_compiler(output):
...
         if "gcc" in command:
-            return _gcc_compiler(output, command)
+            if platform.system() != "Darwin":
+                return _gcc_compiler(output, command)
+
+            # In Mac OS X check if gcc is a fronted using apple-clang
+            _, out = _execute("%s --version" % command)
+            out = out.lower()
+            if "clang" not in out:
+                return _gcc_compiler(output, command)
```
Comment (memsharded) [inline conans/client/conf/detect.py]: The overall logic of this code seems a bit convolute IMHO, with 2 calls to ``_gcc_compiler()`` plus one manual ``_execute("gcc...``. Maybe it would make more sense to put this logic inside ``_gcc_compiler()``, so the compiler is called exactly once, and the rest is processing the output?

File: conans/client/conf/detect.py (hunk 7 above)
Comment (lasote) [reply conans/client/conf/detect.py]: I'm ok with this. If the user is messing with CC/CXX is his problem. And the default profile is just a default.

File: conans/client/conf/detect.py (hunk 8 above)
Comment (pvicente) [reply conans/client/conf/detect.py]: @memsharded @lasote in that case what about the fallback to clang when gcc is a frontend of ``apple-clang``.

Shall we call ```_clang_compiler(output, "clang")``` inside ```_gcc_compiler(output, command)``` as the logic will be inside ``_gcc_compiler``?

File: conans/client/conf/detect.py (hunk 7 above)
Comment (memsharded) [reply conans/client/conf/detect.py]: Ok, makes sense

File: conans/client/conf/detect.py [hunk 9]
```diff
@@ -147,7 +155,15 @@ def _get_default_compiler(output):
...
         command = cc or cxx
         if "gcc" in command:
-            return _gcc_compiler(output, command)
+            gcc = _gcc_compiler(output, command)
+            if platform.system() == "Darwin" and gcc is None:
+                output.warn(
+                    "%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command
+                )
+                # Fallback to use clang instead
+                command = "clang"
```
Comment (memsharded) [inline conans/client/conf/detect.py]: I still think this is a weird outcome. So you defined that your CC=gcc, still conan is detecting and defaulting your profile to clang compiler? I think it is safer to just leave the compiler None for this case, so it will fail fast, instead of lots of potential mismatches between the default profile  and the declared default compiler CC.

Comment (pvicente) [reply conans/client/conf/detect.py]: @memsharded 

>memsharded 5 days ago  Owner
>I am not sure we want to fallback to clang. Because if the env-var CC and CXX are set to gcc, >then, typically CMake will use them when running, and the user might not notice, because it >detected "clang" when CC=gcc. What do you think @lasote?

>  @lasote
>lasote 4 days ago  Owner
>I'm ok with this. If the user is messing with CC/CXX is his problem. And the default profile is >just a default.

>  @memsharded
> memsharded 4 days ago  Owner
> Ok, makes sense

Comment (lasote) [reply conans/client/conf/detect.py]: I talked with @memsharded and I changed my mind, it is better to no set a compiler and make more visible that we are not managing this situation with Conan. Not a big deal.

Comment (pvicente) [reply conans/client/conf/detect.py]: @lasote @memsharded Ok now conan should display:

```
16:13 $ CC=gcc conan install libiconv/1.15@bincrafters/stable
Auto detecting your dev setup to initialize the default profile (/Users/pvicente/.conan/profiles/default)
CC and CXX: gcc, None
ERROR: gcc detected as a frontend using apple-clang. Compiler not supported
ERROR: Unable to find a working compiler
Default settings
	os=Macos
	os_build=Macos
	arch=x86_64
	arch_build=x86_64
	build_type=Release
*** You can change them in /Users/pvicente/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***
libiconv/1.15@bincrafters/stable: Not found in local cache, looking in remotes...
libiconv/1.15@bincrafters/stable: Trying with 'conan-center'...
libiconv/1.15@bincrafters/stable: Trying with 'conan-transit'...
....
```

File: conans/client/conf/detect.py [hunk 10]
```diff
@@ -147,7 +155,12 @@ def _get_default_compiler(output):
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
         if "gcc" in command:
-            return _gcc_compiler(output, command)
+            gcc = _gcc_compiler(output, command)
+            if platform.system() == "Darwin" and gcc is None:
+                output.error(
+                    "%s detected as a frontend using apple-clang. Compiler not supported" % command
```
Comment (memsharded) [inline conans/client/conf/detect.py]: Why not moving the message to ``_gcc_compiler()``? No need to recheck ``platform.system() and gcc is None`` again.
